polar-core/src/vm.rs

```diff
@@ -1406,9 +1406,34 @@ impl PolarVirtualMachine {
             Value::Expression(_) => {
                 return self.query_for_operation(&term);
             }
+            Value::Variable(_a_symbol) => {
+                let val = self.deref(term);
+
+                if val == *term {
+                    // variable was unbound
+                    // apply a constraint to variable that it must be truthy
+                    self.push_goal(Goal::Query {
+                        term: term.clone_with_value(value!(op!(Eq, term.clone(), term!(true)))),
+                    })?;
+                } else {
+                    self.push_goal(Goal::Query { term: val })?;
+                }
+            }
+            Value::Boolean(value) => {
+                if !value {
+                    // Backtrack if the boolean is false.
+                    self.push_goal(Goal::Backtrack)?;
+                }
+
+                return Ok(QueryEvent::None);
+            }
             _ => {
-                let term = self.deref(term);
-                self.query_for_value(&term)?;
+                // everything else dies horribly and in pain
+                let term = self.deref(term); // shadows
```

Comment:
Why do we need to deref here?

Reply:
must attempt to deref to ensure it's unbound

Reply:
We've already checked above for the `Value::Variable` case. There's no harm in doing it again, but it should be redundant.

Hmm... thinking about it, there's maybe a RestVariable case which this might do something with.

Reply:
Nope - sorry - line number shift - this deref is so the error message makes sense, but doh, that's only if it's a variable, and it's not. Getting there slowly